kale/embed/mpca.py

```diff
@@ -72,200 +72,220 @@ class MPCA(BaseEstimator, TransformerMixin):
-                        [proj_mats[m] for m in range(self.ndim - 1) if m != i],
-                        modes=[m for m in range(self.ndim - 1) if m != i],
+                    xj = multi_mode_dot(
+                        x[j, :],  # jth tensor/sample
+                        [proj_mats[m] for m in range(n_dim - 1) if m != i - 1],
+                        modes=[m for m in range(n_dim - 1) if m != i - 1],
                     )
-                    Xj_unfold = unfold(Xj_, i)
-                    Phi[i] = np.dot(Xj_unfold, Xj_unfold.T) + Phi[i]
+                    xj_unfold = unfold(xj, i - 1)
+                    phi[i] = np.dot(xj_unfold, xj_unfold.T) + phi[i]
 
-                eig_vals, eig_vecs = np.linalg.eig(Phi[i])
-                idx_sorted = eig_vals.argsort()[::-1]
-                lambdas[i] = eig_vals[idx_sorted]
-                proj_mats[i] = eig_vecs[:, idx_sorted]
-                proj_mats[i] = (proj_mats[i][:, : shape_out[i]]).T
+                eig_values, eig_vectors = np.linalg.eig(phi[i])
+                idx_sorted = (-1 * eig_values).argsort()
+                # lambdas[i] = eig_values[idx_sorted]
+                proj_mats[i - 1] = (eig_vectors[:, idx_sorted][:, : shape_out[i - 1]]).T
 
-        x_transformed = multi_mode_dot(X, proj_mats, modes=[m for m in range(self.ndim - 1)])
-        x_trans_vec = unfold(x_transformed, mode=-1)  # vectorise the transformed features
+        tensor_pc = multi_mode_dot(x, proj_mats,
+                                   modes=[m for m in range(1, n_dim)])
+        tensor_pc_vec = unfold(tensor_pc, mode=0)  # return_vector the tensor principal components
         # diagonal of the covariance of vectorised features
-        x_trans_cov_diag = np.diag(np.dot(x_trans_vec, x_trans_vec.T))
-        idx_order = x_trans_cov_diag.argsort()[::-1]
+        tensor_pc_cov_diag = np.diag(np.dot(tensor_pc_vec.T, tensor_pc_vec))
+        idx_order = (-1 * tensor_pc_cov_diag).argsort()
 
         self.proj_mats = proj_mats
         self.idx_order = idx_order
         self.shape_out = shape_out
+        self.n_dim = n_dim
 
         return self
 
-    def transform(self, X):
+    def transform(self, x, reduce_mean=True):
         """Perform dimension reduction on X
 
         Args:
-            X (ndarray): Data to perform dimension reduction, shape (I_1, I_2, ..., I_N, n_samples).
+            x (array-like tensor): Data to perform dimension reduction,
+                shape (n_samples, I_1, I_2, ..., I_N).
+            reduce_mean (bool): Whether reduce the mean estimated from training data for X.
+                Defaults to True.
 
         Returns:
-            ndarray: Transformed data, shape (P_1, P_2, ..., P_N, n_samples) if self.vectorise==False.
-                If self.vectorise==True, features will be sorted based on their explained variance ratio,
+            array-like tensor:
+                Transformed data, shape (n_samples, P_1, P_2, ..., P_N) if self.return_vector==False.
```

Comment:
Projected will be better and more precise. Read par 2 of [Wiki PCA](https://en.wikipedia.org/wiki/Principal_component_analysis)